### PR TITLE
Add MultipleImageFetcher

### PR DESCRIPTION
**Pull request by onmyway133**
- Remove `ImageManager`, ping @zenangst 
- Add `MultipleImageFetcher`, this is used to fetch multiple images before hand. This is ideal for prefetching. It uses `ImageFetcher` under the hood

### CODE REVIEW

File: ImaginaryTests-iOS/MultipleImageFetcherTests.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import XCTest
+import Cache
+@testable import Imaginary
+
+fileprivate class MockDownloader: ImageDownloader {
```
Comment (vadymmarkov) [inline ImaginaryTests-iOS/MultipleImageFetcherTests.swift:5]: It could be `private`.

File: ImaginaryTests-iOS/MultipleImageFetcherTests.swift [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  override func download(url: URL, completion: @escaping (Imaginary.Result) -> Void) {
+    self.url = url
+    DispatchQueue.global().async {
+      completion(.value(TestHelper.image()))
+    }
+  }
+}
+
+class MultipleImageFetcherTests: XCTestCase {
+  var storage: Storage!
```
Comment (vadymmarkov) [inline ImaginaryTests-iOS/MultipleImageFetcherTests.swift]: I know it's a test, but could be `private` I think.

File: Sources/Shared/MultipleImageFetcher.swift [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  /// - Parameters:
+  ///   - urls: The urls to fetch
+  ///   - each: Called after each fetcher completed
+  ///   - completion: Called after all fetchers completed
+  public func fetch(urls: [URL],
+                    each: ((Result) -> Void)? = nil,
+                    completion: (([Result]) -> Void)? = nil) {
+    var results: [Result] = []
+
+    self.fetchers = urls.map { url in
```
Comment (vadymmarkov) [inline Sources/Shared/MultipleImageFetcher.swift]: Why do you need this array of fetchers?

File: Sources/Shared/MultipleImageFetcher.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      let fetcher = self.fetcherMaker()
+      fetcher.fetch(url: url, completion: { result in
```
Comment (vadymmarkov) [inline Sources/Shared/MultipleImageFetcher.swift]: I wouldn't do `fetch` inside `map`, it should be used for transformation.

**Review by vadymmarkov — COMMENTED** [review]
@onmyway133 Why do we need this `MultipleImageFetcher`? I see that we can load multiple images, but do you have a real use case when you would need it?

File: ImaginaryTests-iOS/MultipleImageFetcherTests.swift (hunk 2 above)
Comment (onmyway133) [reply ImaginaryTests-iOS/MultipleImageFetcherTests.swift]: 👍

File: Sources/Shared/MultipleImageFetcher.swift (hunk 3 above)
Comment (onmyway133) [reply Sources/Shared/MultipleImageFetcher.swift]: I want to keep the references, to make them alive

File: Sources/Shared/MultipleImageFetcher.swift (hunk 4 above)
Comment (onmyway133) [reply Sources/Shared/MultipleImageFetcher.swift]: @vadymmarkov I will fix this, agree

File: Sources/Shared/MultipleImageFetcher.swift (hunk 3 above)
Comment (zenangst) [reply Sources/Shared/MultipleImageFetcher.swift]: If you don't want them to be alive anymore, don't forget to burn them. #popculturereferences 📺

**Comment by onmyway133** [comment]
@vadymmarkov I fixed, may give it another look ?

File: Sources/Shared/MultipleImageFetcher.swift [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  public func fetch(urls: [URL],
+                    each: ((Result) -> Void)? = nil,
+                    completion: (([Result]) -> Void)? = nil) {
+    self.fetchers = urls.map { _ in
+      return self.fetcherMaker()
+    }
+
+    var results: [Result] = []
+
+    zip(fetchers, urls).forEach { fetcher, url in
```
Comment (vadymmarkov) [inline Sources/Shared/MultipleImageFetcher.swift:28]: @onmyway133 You couldn't do it without functional programming, didn't you? 😄

**Review by vadymmarkov — APPROVED** [review]

File: Sources/Shared/MultipleImageFetcher.swift (hunk 5 above)
Comment (onmyway133) [reply Sources/Shared/MultipleImageFetcher.swift:28]: @vadymmarkov 😛

**Pull request merged**